kong/runloop/plugins_iterator.lua

```diff
@@ -123,110 +163,176 @@ local function get_next(self)
-        if plugin_configuration then
-          break
-        end
+    if route_id and service_id and combos[COMBO_RS] then
+      plugin_configuration = load_configuration(ctx, name, route_id, service_id, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if route_id and service_id and combos[COMBO_RS] then
-        plugin_configuration = load_configuration(self, name, route_id, service_id, nil)
-        if plugin_configuration then
-          break
-        end
+    if consumer_id and combos[COMBO_C] then
+      plugin_configuration = load_configuration(ctx, name, nil, nil, consumer_id)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if consumer_id and combos[COMBO_C] then
-        plugin_configuration = load_configuration(self, name, nil, nil, consumer_id)
-        if plugin_configuration then
-          break
-        end
-      end
-
-      if route_id and combos[COMBO_R] then
-        plugin_configuration = load_configuration(self, name, route_id, nil, nil)
-        if plugin_configuration then
-          break
-        end
+    if route_id and combos[COMBO_R] then
+      plugin_configuration = load_configuration(ctx, name, route_id, nil, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if service_id and combos[COMBO_S] then
-        plugin_configuration = load_configuration(self, name, nil, service_id, nil)
-        if plugin_configuration then
-          break
-        end
+    if service_id and combos[COMBO_S] then
+      plugin_configuration = load_configuration(ctx, name, nil, service_id, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if combos[COMBO_GLOBAL] then
-        plugin_configuration = load_configuration(self, name, nil, nil, nil)
-      end
+    if combos[COMBO_GLOBAL] then
+      plugin_configuration = load_configuration(ctx, name, nil, nil, nil)
+    end
 
-    until true
+    ::after_combos::
 
     if plugin_configuration then
       ctx.plugins[name] = plugin_configuration
     end
   end
 
+  local phase = self.iterator.phases[self.phase]
   -- return the plugin configuration
-  if ctx.plugins[plugin.name] then
+  if phase and phase[plugin.name] and ctx.plugins[plugin.name] then
     return plugin, ctx.plugins[plugin.name]
   end
 
   return get_next(self) -- Load next plugin
 end
 
 
-local plugins_iterator_mt = { __call = get_next }
+local PluginsIterator = {}
 
 
 --- Plugins for request iterator.
 -- Iterate over the plugin loaded for a request, stored in
 --`ngx.ctx.plugins`.
 --
--- @param[type=table] ctx Nginx context table
 -- @param[type=string] phase Plugins iterator execution phase
--- @param[type=table] plugins Plugins table
 -- @treturn function iterator
-local function plugins_iterator(ctx, phase, plugins_plan)
+local function iterate(self, phase)
+  local ctx = ngx.ctx
```

Comment:
I think the `ctx` should be passed here as optimization to avoid `ngx.ctx`.

Reply:
done